lib/fix_ints.nit

```diff
@@ -0,0 +1,789 @@
+# distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without  even  the implied warranty of  MERCHANTABILITY or  FITNESS FOR A
+# PARTICULAR PURPOSE.  You can modify it is you want,  provided this header
+# is kept unaltered, and a notification of the changes is added.
+# You  are  allowed  to  redistribute it and sell it, alone or is a part of
+# another product.
+
+# Integers of fixed-size with C-like semantics
+#
+# Since Nit vows to make life easier for higher-level programming
+# It lacked a bit in low-level int representation
+#
+# This module introduces several fixed-length integers representations
+#
+# You may use it for special occasions when dealing with close-to-the-metal situations
+#
+# Specifically for:
+#
+# * Cryptography-related operations
+# * Virtual Machine implementation
+# * Emulators
+# * etc.
+#
+# All representations included here are akin to C's stdint defined types
+# When using these, bear in mind that calculations on different types here may overflow
+#
+# Also, when converting from one type to another, unless expected, keep in mind that an overflow
+# may happen, and that by doing so, some data may be lost in translation.
+module fix_ints
+
+`{
+	#include <stdint.h>
+`}
+
+redef class Numeric
+	# Returns `self` as `int8_t`
+	fun to_i8: Int8 is abstract
+	# Returns `self` as `int16_t`
+	fun to_i16: Int16 is abstract
+	# Returns `self` as `uint16_t`
+	fun to_u16: UInt16 is abstract
+	# Returns `self` as `int32_t`
+	fun to_i32: Int32 is abstract
+	# Returns `self` as `uint32_t`
+	fun to_u32: UInt32 is abstract
+end
+
+# Any kind of native int type with a fixed-size
+interface NativeInt
+	super Numeric
+	super Discrete
+
+	redef type OTHER: NativeInt
+
+	# Minimum value of `self`
+	fun min_int: Int is abstract
+	# Maximum value of `self`
+	fun max_int: Int is abstract
+
+	# Returns `self` rshifted by `o` bits
+	fun >>(o: Int): SELF is abstract
+	# Returns `self` lshifted by `o` bits
+	fun <<(o: Int): SELF is abstract
+	# Binary OR of `self` and `o`
+	fun |(o: SELF): SELF is abstract
+	# Binary AND of `self` and `o`
+	fun &(o: SELF): SELF is abstract
+	# Binary NOT of `self`
+	fun ~: SELF is abstract
+	# Binary XOR of `self` and `o`
+	fun ^(o: SELF): SELF is abstract
+	# Modulo of `self` with `o`.
+	#
+	# Finds the remainder of division of `self` by `o`.
+	fun %(o: SELF): SELF is abstract
+
+	# C function to calculate the length of the `NativeString` to receive `self`
+	private fun strlen: Int is abstract
+
+	# C function to convert an nit Int16 to a NativeString (char*)
+	private fun native_to_s(nstr: NativeString, strlen: Int) is abstract
+
+	redef fun to_s do
+		var nslen = strlen
+		var ns = new NativeString(nslen + 1)
+		ns[nslen] = 0u8
+		native_to_s(ns, nslen + 1)
+		return ns.to_s_with_length(nslen)
+	end
+end
+
+# Signed integer with 8-bit of precision
+#
+# Max int = +127
+# Min int = -128
+extern class Int8 `{ int8_t * `}
```

Comment:
Is the pointer type really necessary? You should be able to directly use the int as the scalar extern type, same for the other types which are the same size or smaller than a `long`. To make if official, we could force the size of the `val*` structure to accept extern types up to 64 bits and even have a new root object distinct of Pointer.

...but it should probably be intern/universal anyway.